From 0f8ecb74dcdf93af60bf2c727575febdc32ac30b Mon Sep 17 00:00:00 2001
From: Anthoine Bourgeois <anthoine.bourgeois@vates.tech>
Date: Thu, 16 Jan 2025 14:09:16 +0100
Subject: [PATCH] libqcow2: manage qcow2 sources

This commit includes a list of files imported from qemu to build
the qcow2 library for tapdisk.
A small script helps to import files and diff them to ease the
sources management.

Signed-off-by: Anthoine Bourgeois <anthoine.bourgeois@vates.tech>
---
 manage-qemu-sources.sh |  74 +++++++++++++
 qemu-files.lst         | 246 +++++++++++++++++++++++++++++++++++++++++
 2 files changed, 320 insertions(+)
 create mode 100755 manage-qemu-sources.sh
 create mode 100644 qemu-files.lst

diff --git a/manage-qemu-sources.sh b/manage-qemu-sources.sh
new file mode 100755
index 00000000..bba47d84
--- /dev/null
+++ b/manage-qemu-sources.sh
@@ -0,0 +1,74 @@
+#!/bin/bash
+
+usage_function () {
+    echo "Usage: $0 qemu_directory [diff|sync]"
+    exit 0
+}
+
+if [ $# -ne 2 ]; then
+    usage_function
+fi
+
+qemudir=${1}
+cmd=${2}
+
+if [ ${cmd} != "diff" -a ${cmd} != "sync" ]; then
+    echo "Wring command: choose 'diff' or 'sync'"
+    usage_function
+fi
+
+diff_function () {
+    if [ ! -f ${qemudir}/${src_prefix}/${1} ]; then
+        echo "${qemudir}/${src_prefix}/${1} doesn't exist."
+        exit 0
+    fi
+    if [ ! -f ${dst_prefix}/${1} ]; then
+        echo "${dst_prefix}/${1} doesn't exist."
+        exit 0
+    fi
+    diff -q ${qemudir}/${src_prefix}/${1} ${dst_prefix}/${1}
+    if [ $? -ne 0 ]; then
+        diff -Npur ${qemudir}/${src_prefix}/${1} ${dst_prefix}/${1}
+    fi
+}
+
+sync_warning () {
+    echo "This command will ease all the modifications you did"
+    echo "in all the qemu files you imported."
+    echo -n "Are you sure you want to sync all the files ? (N/y) "
+    read ok
+    if [ "$ok" == "y" ]; then
+        echo "Fine! Start in:"
+        for i in `seq 5`; do
+            echo -n "$((6 - i)) "
+            sleep 1
+        done
+        echo "Let's go!"
+    else
+        echo "Abort!"
+        exit 0
+    fi
+}
+
+sync_function () {
+    echo "cp ${qemudir}/${src_prefix}/${1} ${dst_prefix}/${1}"
+    mkdir -p $(dirname "${dst_prefix}/${1}")
+    cp -f "${qemudir}/${src_prefix}/${1}" "${dst_prefix}/${1}"
+}
+
+if [ ${cmd} == "sync" ]; then
+    sync_warning
+fi
+
+while read f; do
+    if [ ${f:0:1} == '#' ]; then
+        src_prefix=`echo ${f:1} | awk -F ':' '{print $1}'`
+        dst_prefix=`echo ${f:1} | awk -F ':' '{print $2}'`
+        continue
+    fi
+    if [ ${cmd} == "diff" ]; then
+        diff_function ${f} ${src_prefix} ${dst_prefix}
+    elif [ ${cmd} == "sync" ]; then
+        sync_function ${f} ${src_prefix} ${dst_prefix}
+    fi
+done < qemu-files.lst
diff --git a/qemu-files.lst b/qemu-files.lst
new file mode 100644
index 00000000..a9a7e4d2
--- /dev/null
+++ b/qemu-files.lst
@@ -0,0 +1,246 @@
+#.:.
+include/block/accounting.h
+include/block/aio.h
+include/block/aio-wait.h
+include/block/aio_task.h
+include/block/block-common.h
+include/block/block-global-state.h
+include/block/block-io.h
+include/block/block.h
+include/block/block_int-common.h
+include/block/block_int-global-state.h
+include/block/block_int-io.h
+include/block/block_int.h
+include/block/blockjob.h
+include/block/blockjob_int.h
+include/block/dirty-bitmap.h
+include/block/graph-lock.h
+include/block/qapi.h
+include/block/qdict.h
+include/block/raw-aio.h
+include/block/snapshot.h
+include/block/thread-pool.h
+include/crypto/hash.h
+include/exec/hwaddr.h
+include/glib-compat.h
+include/hw/block/block.h
+include/qapi/compat-policy.h
+include/qapi/dealloc-visitor.h
+include/qapi/error.h
+include/qapi/qmp-event.h
+include/qapi/qmp/dispatch.h
+include/qapi/qmp/json-parser.h
+include/qapi/qmp/json-writer.h
+include/qapi/qmp/qbool.h
+include/qapi/qmp/qdict.h
+include/qapi/qmp/qerror.h
+include/qapi/qmp/qjson.h
+include/qapi/qmp/qlist.h
+include/qapi/qmp/qnull.h
+include/qapi/qmp/qnum.h
+include/qapi/qmp/qobject.h
+include/qapi/qmp/qstring.h
+include/qapi/qobject-input-visitor.h
+include/qapi/qobject-output-visitor.h
+include/qapi/util.h
+include/qapi/visitor-impl.h
+include/qapi/visitor.h
+include/qemu/atomic.h
+include/qemu/bitmap.h
+include/qemu/bitops.h
+include/qemu/bswap.h
+include/qemu/clang-tsa.h
+include/qemu/compiler.h
+include/qemu/config-file.h
+include/qemu/coroutine-core.h
+include/qemu/coroutine-tls.h
+include/qemu/coroutine.h
+include/qemu/coroutine_int.h
+include/qemu/ctype.h
+include/qemu/cutils.h
+include/qemu/defer-call.h
+include/qemu/error-report.h
+include/qemu/event_notifier.h
+include/qemu/futex.h
+include/qemu/hbitmap.h
+include/qemu/help_option.h
+include/qemu/host-utils.h
+include/qemu/id.h
+include/qemu/iov.h
+include/qemu/job.h
+include/qemu/lockable.h
+include/qemu/main-loop.h
+include/qemu/memalign.h
+include/qemu/module.h
+include/qemu/notify.h
+include/qemu/option.h
+include/qemu/option_int.h
+include/qemu/osdep.h
+include/qemu/processor.h
+include/qemu/progress_meter.h
+include/qemu/qdist.h
+include/qemu/qemu-print.h
+include/qemu/qemu-progress.h
+include/qemu/qsp.h
+include/qemu/queue.h
+include/qemu/range.h
+include/qemu/ratelimit.h
+include/qemu/rcu.h
+include/qemu/rcu_queue.h
+include/qemu/stats64.h
+include/qemu/sys_membarrier.h
+include/qemu/thread-posix.h
+include/qemu/thread.h
+include/qemu/timed-average.h
+include/qemu/timer.h
+include/qemu/transactions.h
+include/qemu/typedefs.h
+include/qemu/unicode.h
+include/qemu/units.h
+include/sysemu/block-backend-common.h
+include/sysemu/block-backend-global-state.h
+include/sysemu/block-backend-io.h
+include/sysemu/block-backend.h
+include/sysemu/blockdev.h
+include/sysemu/cpu-timers.h
+include/sysemu/os-posix.h
+#block:include
+qcow2.h
+#build:include
+block/module_block.h
+qapi/qapi-builtin-types.h
+qapi/qapi-builtin-visit.h
+qapi/qapi-commands-block.h
+qapi/qapi-commands-block-core.h
+qapi/qapi-commands-common.h
+qapi/qapi-commands-job.h
+qapi/qapi-events-block-core.h
+qapi/qapi-events-common.h
+qapi/qapi-events-job.h
+qapi/qapi-types-block.h
+qapi/qapi-types-block-core.h
+qapi/qapi-types-common.h
+qapi/qapi-types-compat.h
+qapi/qapi-types-error.h
+qapi/qapi-types-job.h
+qapi/qapi-types-run-state.h
+qapi/qapi-visit-block-core.h
+qapi/qapi-visit-common.h
+qapi/qapi-visit-job.h
+#.:qcow2/lib
+block.c
+blockdev.c
+blockjob.c
+block/accounting.c
+block/aio_task.c
+block/block-backend.c
+block/block-gen.h
+block/commit.c
+block/coroutines.h
+block/dirty-bitmap.c
+block/file-posix.c
+block/graph-lock.c
+block/io.c
+block/linux-aio.c
+block/mirror.c
+block/monitor/bitmap-qmp-cmds.c
+block/progress_meter.c
+block/qapi.c
+block/raw-format.c
+block/snapshot.c
+crypto/hash.c
+hw/block/block.c
+hw/block/hd-geometry.c
+job.c
+job-qmp.c
+qapi/qapi-dealloc-visitor.c
+qapi/qapi-util.c
+qapi/qapi-visit-core.c
+qapi/qobject-input-visitor.c
+qapi/qobject-output-visitor.c
+qobject/block-qdict.c
+qobject/json-lexer.c
+qobject/json-parser.c
+qobject/json-parser-int.h
+qobject/json-streamer.c
+qobject/json-writer.c
+qobject/qbool.c
+qobject/qdict.c
+qobject/qjson.c
+qobject/qlist.c
+qobject/qnull.c
+qobject/qnum.c
+qobject/qobject-internal.h
+qobject/qobject.c
+qobject/qstring.c
+system/cpus.c
+util/aio-posix.c
+util/aio-posix.h
+util/aio-wait.c
+util/aiocb.c
+util/async.c
+util/bitmap.c
+util/bitops.c
+util/bufferiszero.c
+util/compatfd.c
+util/coroutine-ucontext.c
+util/cutils.c
+util/defer-call.c
+util/error-report.c
+util/error.c
+util/event_notifier-posix.c
+util/fdmon-epoll.c
+util/fdmon-poll.c
+util/hbitmap.c
+util/host-utils.c
+util/id.c
+util/iov.c
+util/lockcnt.c
+util/main-loop.c
+util/memalign.c
+util/module.c
+util/notify.c
+util/osdep.c
+util/oslib-posix.c
+util/qemu-coroutine-lock.c
+util/qemu-coroutine.c
+util/qemu-option.c
+util/qemu-print.c
+util/qemu-thread-common.h
+util/qemu-thread-posix.c
+util/qemu-timer.c
+util/qsp.c
+util/rcu.c
+util/stats64.c
+util/thread-pool.c
+util/timed-average.c
+util/transactions.c
+util/unicode.c
+qapi/qmp-dispatch.c
+qapi/qmp-event.c
+#block:qcow2/lib
+qcow2-bitmap.c
+qcow2-cache.c
+qcow2-cluster.c
+qcow2-refcount.c
+qcow2-snapshot.c
+qcow2-threads.c
+qcow2.c
+#build:qcow2/lib
+block/block-gen.c
+qapi/qapi-emit-events.h
+qapi/qapi-events-block-core.c
+qapi/qapi-events-job.c
+qapi/qapi-types-block-core.c
+qapi/qapi-types-common.c
+qapi/qapi-types-job.c
+qapi/qapi-visit-block-core.c
+qapi/qapi-visit-common.c
+qapi/qapi-visit-job.c
+#.:qcow2
+qemu-img.c
+#util:qcow2
+qemu-config.c
+qemu-progress.c
+#build:qcow2
+qemu-img-cmds.h
